src/Rules/RuleLevelHelper.php

```diff
@@ -76,6 +78,17 @@ public function accepts(Type $acceptingType, Type $acceptedType, bool $strictTyp
 		}
 
 		if ($acceptingType instanceof UnionType && !$acceptedType instanceof CompoundType) {
+			if (
```

Comment:
Hi, I don't see why RuleLevelHelper would have to be changed for this.

Reply:
Don't know why but there is an early check here:

When I check `\DateTimeInterface|int` vs `\DateTime|\DateTimeImmutable|int` it's a compound type, so this is not done here and ends up in the UnionType::accepts methods.
But when I check `\DateTimeInterface` vs `\DateTime|\DateTimeImmutable` there is an early check here which return either true or false, so the UnionType::accepts method is never executed.

Reply:
The issue is about `DateTime|DateTimeImmutable` accepting `DateTimeInterface`. So only UnionType::accepts() should be modified.

Reply:
acceptingType = DateTime|DateTimeImmutable
acceptedType = DateTimeInterface

```
if ($acceptingType instanceof UnionType && !$acceptedType instanceof CompoundType) {
```
We're in this if.

So we're currently doing
```
foreach ($acceptingType->getTypes() as $innerType) {
     if (self::accepts($innerType, $acceptedType, $strictTypes)) {
          return true;
     }
 }
return false;
```
and since neither DateTime doesn't accept DateTimeInterface, neither DateTimeImmutable accept DateTimeInterface, false is returned.

I can change the check 
```
if ($acceptingType instanceof UnionType && !$acceptedType instanceof CompoundType) {
```
to something like
```
if ($acceptingType instanceof UnionType && !$acceptedType instanceof CompoundType && !$acceptedType->equals(new ObjectType(\DateTimeInterface::class))) {
```
if you prefer, but something has to be done here.

If I remove the code added here, tests are failing.

Reply:
WDYT @ondrejmirtes ?

Reply:
The change in RuleLevelHelper shouldn't be specific to the DateTime situation, but more general. I propose something like this:

```diff
diff --git a/src/Rules/RuleLevelHelper.php b/src/Rules/RuleLevelHelper.php
index 240258921..7fc9724b8 100644
--- a/src/Rules/RuleLevelHelper.php
+++ b/src/Rules/RuleLevelHelper.php
@@ -75,7 +75,8 @@ class RuleLevelHelper
 			$acceptedType = TypeCombinator::removeNull($acceptedType);
 		}
 
-		if ($acceptingType instanceof UnionType && !$acceptedType instanceof CompoundType) {
+		$accepts = $acceptingType->accepts($acceptedType, $strictTypes);
+		if (!$accepts->yes() && $acceptingType instanceof UnionType && !$acceptedType instanceof CompoundType) {
 			foreach ($acceptingType->getTypes() as $innerType) {
 				if (self::accepts($innerType, $acceptedType, $strictTypes)) {
 					return true;
@@ -103,8 +104,6 @@ class RuleLevelHelper
 			);
 		}
 
-		$accepts = $acceptingType->accepts($acceptedType, $strictTypes);
-
 		return $this->checkUnionTypes ? $accepts->yes() : !$accepts->no();
 	}
 

```

Reply:
Yeah I didn't like to be specific to the DateTime but It wasn't clear to me what to do.

I used your proposal and seems to work perfectly ; thanks :)